h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -437,6 +440,123 @@ public void modelFailed(Model failedModel) {
     }
   } // class CartesianWalker
 
+
+  /**
+   * Hyperparameter space walker which visits each combination of hyperparameters in order.
+   */
+  public static class MaterializedRandomWalker<MP extends Model.Parameters>
+          extends BaseWalker<MP, HyperSpaceSearchCriteria.RandomDiscreteValueSearchCriteria> {
+
+    public MaterializedRandomWalker(MP params,
+                                    Map<String, Object[]> hyperParams,
+                                    ModelParametersBuilderFactory<MP> paramsBuilderFactory,
+                                    HyperSpaceSearchCriteria.RandomDiscreteValueSearchCriteria search_criteria,
+                                    ArrayList<Function<HashMap<String, Object>, HashMap<String, Object>>> filterFunctions) {
+      super(params, hyperParams, paramsBuilderFactory, search_criteria);
+    }
+
+    @Override
+    public HyperSpaceIterator<MP> iterator() {
+      ArrayList<HashMap<String, Object>> allElements = new ArrayList<>();
+
+      _hyperParams.forEach((hpKey, hpValues) -> {
+        if(allElements.isEmpty()) {
+          Arrays.stream(hpValues).forEach(hpValue -> {
+            HashMap<String, Object> newGridItem = new HashMap<>();
+            newGridItem.put(hpKey, hpValue);
+            allElements.add(newGridItem);
+          });
+        } else {
+          Stream<HashMap<String, Object>> expandedResult = allElements.stream().flatMap(existingItem -> {
+            return Arrays.stream(hpValues).map(hpValue -> {
+              HashMap<String, Object> clone = (HashMap<String, Object>) existingItem.clone();
+              clone.put(hpKey, hpValue);
+              return clone;
+            });
+          });
+          ArrayList<HashMap<String, Object>> collect = expandedResult.collect(Collectors.toCollection(ArrayList::new));
+          allElements.clear();
+          allElements.addAll(collect);
+        }
+      });
+      Collections.shuffle(allElements);
```

Comment:
All operations involving a random number generator need to be seeded. Grid search should be reproducible.

Reply:
@michalkurka yes, definitely it will be added.